### PR TITLE
feat: replace Schema with FieldLayout for contract internals

### PR DESCRIPTION
**Pull request by dk1a**
fixes https://github.com/latticexyz/mud/issues/1326
see the changeset for more context on this change

### CODE REVIEW

**Comment by alvrs** [comment]
High level feedback on terminology: I think we should keep calling the array of schema types "schema", which is the thing that is used by off-chain consumers to know how to interpret the data, and we should find a new name for the "static byte lengths" object used internally by `Store` and the table libraries. Otherwise it gets confusing if we use "Schema" for one thing onchain and another thing offchain.

**Comment by dk1a** [comment]
> I think we should keep calling the array of schema types "schema"

Technically no more schema types, just abi types (I wanna ditch the enum and replace it with `bytes32` strings that directly mirror config/client).
They can still be called Schema. Or we could rename to something like `AbiTypes`. `getValueAbiTypes` `getKeyAbiTypes`? (doesn't sound as good I guess)

> we should find a new name for the "static byte lengths"

Agree, `FieldMetadata` `FieldInfo`? This stores the number of fields and their lengths, so `FieldLengths` wouldn't really work.

File: packages/store/src/StoreCore.sol [hunk 1]
```diff
@@ -16,14 +16,14 @@ import { StoreSwitch } from "./StoreSwitch.sol";
...
-  event StoreSetField(bytes32 table, bytes32[] key, uint8 schemaIndex, bytes data);
+  event StoreSetField(bytes32 table, bytes32[] key, uint8 fieldIndex, bytes data);
   event StoreDeleteRecord(bytes32 table, bytes32[] key);
   event StoreEphemeralRecord(bytes32 table, bytes32[] key, bytes data);
 
   /**
    * Initialize internal tables.
    * Consumers must call this function in their constructor.
-   * TODO: should we turn the schema table into a "proper table" and register it here?
+   * TODO: should we turn the FieldLayout table into a "proper table" and register it here?
```
Comment (alvrs) [inline packages/store/src/StoreCore.sol:26]: this TODO is actually outdated, we can remove it

File: packages/store/ts/codegen/renderTable.ts [hunk 2]
```diff
@@ -104,9 +112,9 @@ library ${libraryName} {
   ${renderWithStore(
     storeArgument,
     (_typedStore, _store, _commentSuffix) => `
-    /** Register the table's key schema, value schema, key names and value names${_commentSuffix} */
+    /** Register the table keys' and values' field layout, key names and value names${_commentSuffix} */
```
Comment (holic) [inline packages/store/ts/codegen/renderTable.ts:115]: I don't think field layout applies to keys, so I went to adjust this comment, but then realized maybe we don't need to include all of this method's implementation in the comment? Maybe just calling out "its config" is enough?

```suggestion
    /** Register the table with its config${_commentSuffix} */
```

File: .changeset/grumpy-files-heal.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+---
+"@latticexyz/cli": major
+"@latticexyz/protocol-parser": minor
+"@latticexyz/store": major
+"@latticexyz/world": major
+---
+
+- Add `FieldLayout`, which is similar to `Schema`, but with static byte lengths instead of `SchemaType`s
```
Comment (holic) [inline .changeset/grumpy-files-heal.md:8]: Since this is a major change, can we add more info here? Explain what is/was `Schema`, what is `FieldLayout`, and why we're making this change?

File: packages/world/gas-report.json
Comment (alvrs) [inline packages/world/gas-report.json:1]: great to see all these savings 🔥

**Review by alvrs — COMMENTED** [review]
looks clean overall and great gas savings!

File: packages/store/src/StoreCore.sol [hunk 4]
```diff
@@ -64,49 +64,67 @@ library StoreCore {
...
+   * Register a new table with key field layout, value field layout, key names, and value names
    */
   function registerTable(
     bytes32 tableId,
+    FieldLayout fieldLayout,
     Schema keySchema,
     Schema valueSchema,
     string[] memory keyNames,
-    string[] memory fieldNames
+    string[] memory valueNames
```
[the remark this answers is not in the document]
Comment (holic) [reply packages/store/src/StoreCore.sol:93]: for context, we went with "field names" instead of "value names" because we didn't have a good solution for the name of values ("field values" since "value values" sounds weird)

File: packages/cli/src/utils/deploy.ts [hunk 5]
```diff
@@ -158,6 +158,12 @@ export async function deploy(
         return schemaType;
       });
 
+      const schemaTypeLengths = schemaTypes.map((schemaType) => getStaticByteLength(schemaType));
+      const fieldLayout = {
+        staticFieldLengths: schemaTypeLengths.filter((schemaTypeLength) => schemaTypeLength > 0),
+        numDynamicFields: schemaTypeLengths.filter((schemaTypeLength) => schemaTypeLength === 0).length,
+      };
+
```
[the remark this answers is not in the document]
Comment (holic) [reply packages/cli/src/utils/deploy.ts:166]: I spotted this too, I can take migrating this.

File: packages/store/src/IStoreErrors.sol [hunk 6]
```diff
@@ -10,6 +10,7 @@ interface IStoreErrors {
   error StoreCore_NotDynamicField();
   error StoreCore_InvalidDataLength(uint256 expected, uint256 received);
   error StoreCore_InvalidKeyNamesLength(uint256 expected, uint256 received);
-  error StoreCore_InvalidFieldNamesLength(uint256 expected, uint256 received);
+  error StoreCore_InvalidValueNamesLength(uint256 expected, uint256 received);
```
[the remark this answers is not in the document]
Comment (dk1a) [reply packages/store/src/IStoreErrors.sol:13]: But we call a table column `field`, whereas `key` and `value` are used for the 2 respective things that have schemas
I think my point is that you access a `field` via `keys` to get its `value`. So `value` seems more informative in this case.
`FieldNames` also makes sense I suppose, it's ur call

File: packages/store/ts/codegen/renderTable.ts [hunk 7]
```diff
@@ -73,6 +73,14 @@ ${
 }
 
 library ${libraryName} {
+  /** Get the table values' field layout */
+  function getFieldLayout() internal pure returns (FieldLayout) {
+    uint256[] memory _fieldLayout = new uint256[](${staticFields.length});
+    ${renderList(staticFields, ({ staticByteLength }, index) => `_fieldLayout[${index}] = ${staticByteLength};`)}
+
+    return FieldLayoutLib.encode(_fieldLayout, ${dynamicFields.length});
+  }
```
[the remark this answers is not in the document]
Comment (dk1a) [reply packages/store/ts/codegen/renderTable.ts:82]: yep

File: packages/store/src/FieldLayout.sol [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+      } else if (staticByteLength > 32) {
+        revert FieldLayoutLib_StaticTypeDoesNotFitInAWord();
+      }
+
+      unchecked {
+        // (safe because 28 (max _staticFields.length) * 32 (max static length) < 2**16)
+        totalLength += staticByteLength;
+        // Sequentially store lengths after the first 4 bytes (which are reserved for total length and field numbers)
+        // (safe because of the initial _staticFields.length check)
+        fieldLayout |= uint256(_staticFields[i]) << ((31 - 4 - i) * 8);
```
[the remark this answers is not in the document]
Comment (dk1a) [reply packages/store/src/FieldLayout.sol]: added some constants, the `31` was confusing and should only really be in 1 place, I use a special constant there

[the remark this answers is not in the document]
Comment (dk1a) [reply packages/store/src/FieldLayout.sol]: btw don't mind the gas increase, it was misleading and probably wrong to begin with, see https://github.com/latticexyz/mud/issues/1267

File: packages/store/test/FieldLayout.t.sol [hunk 9]
```diff
@@ -0,0 +1,207 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+import { Test, console } from "forge-std/Test.sol";
+import { GasReporter } from "@latticexyz/gas-report/src/GasReporter.sol";
+import { FieldLayout, FieldLayoutLib } from "../src/FieldLayout.sol";
+import { FieldLayoutEncodeHelper } from "./FieldLayoutEncodeHelper.sol";
+
+// TODO add tests for all lengths
```
[the remark this answers is not in the document]
Comment (dk1a) [reply packages/store/test/FieldLayout.t.sol:9]: it's meant in the same sense as `// TODO add tests for all schema types` in `Schema.t.sol`
we could fuzz/autogenerate tests for all the length configurations (or at least a lot more combinations), rather than testing a few edge cases

afaik we don't have an issue, made https://github.com/latticexyz/mud/issues/1453

File: packages/store/src/FieldLayout.sol [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+/**
+ * Static functions for FieldLayout
+ */
+library FieldLayoutLib {
+  error FieldLayoutLib_InvalidLength(uint256 length);
+  error FieldLayoutLib_StaticTypeIsZero();
+  error FieldLayoutLib_StaticTypeDoesNotFitInAWord();
+
+  // Based on PackedCounter's capacity
+  uint256 internal constant MAX_DYNAMIC_FIELDS = 5;
```
[the remark this answers is not in the document]
Comment (dk1a) [reply packages/store/src/FieldLayout.sol]: added a constants file and also refactored Schema to use it

File: .changeset/grumpy-files-heal.md (hunk 3 above)
Comment (dk1a) [reply .changeset/grumpy-files-heal.md]: added some more info, not sure how far to go, do tell if u want even more details

File: .changeset/grumpy-files-heal.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+`FieldLayout` is more gas-efficient because it already has lengths, and `Schema` has types which need to be converted to lengths.
+
+3. Add `getFieldLayout` to `IStore` interface.
+
+There is no `FieldLayout` for keys, only for values, because key byte lengths aren't usually relevant on-chain.
+You can still use `getKeySchema` if you need key types.
+
+4. Add `fieldLayoutToHex` utility to `protocol-parser` package.
+
+5. Add `constants.sol` for constants shared between `FieldLayout`, `Schema` and `PackedCounter`.
```
Comment (holic) [inline .changeset/grumpy-files-heal.md:30]:
```suggestion
- Add `FieldLayout`, which is a `bytes32` user-type similar to `Schema`.

   Both `FieldLayout` and `Schema` have the same kind of data in the first 4 bytes.

   - 2 bytes for total length of all static fields
   - 1 byte for number of static size fields
   - 1 byte for number of dynamic size fields

   But whereas `Schema` has `SchemaType` enum in each of the other 28 bytes, `FieldLayout` has static byte lengths in each of the other 28 bytes.

- Replace `Schema valueSchema` with `FieldLayout fieldLayout` in Store and World contracts.

   `FieldLayout` is more gas-efficient because it already has lengths, and `Schema` has types which need to be converted to lengths.

- Add `getFieldLayout` to `IStore` interface.

   There is no `FieldLayout` for keys, only for values, because key byte lengths aren't usually relevant on-chain. You can still use `getKeySchema` if you need key types.

- Add `fieldLayoutToHex` utility to `protocol-parser` package.

- Add `constants.sol` for constants shared between `FieldLayout`, `Schema` and `PackedCounter`.
```

nit: moved bullets over to an unordered list and indented the content of each bullet so that it renders properly

Keep in mind that each package in the list at the top will get this exact entry added to the changelog. We might consider adding separate changeset file per package to note package-specific changes (e.g. `constants.sol` only applies to world package)

**Review by alvrs — DISMISSED** [review]
code looks good! good to be merged once merge conflicts are resolved

**Review by alvrs — APPROVED** [review]

File: e2e/packages/contracts/types/ethers-contracts/IWorld.ts [hunk 12]
```diff
@@ -0,0 +1,1734 @@
+/* Autogenerated file. Do not edit manually. */
+/* tslint:disable */
+/* eslint-disable */
```
Comment (holic) [inline e2e/packages/contracts/types/ethers-contracts/IWorld.ts:3]: these files snuck back in 😭

File: packages/store/gas-report.json [hunk 13]
```diff
@@ -459,211 +519,217 @@
     "file": "test/StoreCoreDynamic.t.sol",
     "test": "testGetFieldSlice",
     "name": "get field slice (cold, 1 slot)",
-    "gasUsed": 7990
+    "gasUsed": 8425
```
[the remark this answers is not in the document]
Comment (dk1a) [reply packages/store/gas-report.json:522]: dunno, I assume it's #1267, but could be the optimizer being bad too
